### PR TITLE
Add serialize column to civicrm_custom_field, remove "Multi-Select" html type

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Part of an effort to make custom fields more like core fields, this adds a "serialize" attribute (which core fields have) and relies on that attribute to determine if a Select field is multivalued. Previously custom fields would rename the `html_type` from "Select" to "Multi-Select" to achieve that distinction, which led to a lot of odd stuff.

Before
----------------------------------------
Impossible to tell how custom data was serialized without guessing based on the `html_type`.

After
----------------------------------------
Serialization declared as part of the field metadata.
The following field types have been removed and existing fields updated to include `serialize = 1` and their types changed to:
- Multi-Select => Select
- Multi-Select Country => Select Country
- Multi-Select State/Province => Select State/Province

Technical Details
-------------
A compatibility layer has been added to APIv3. *By default* APIv3 operates in legacy mode, changing input/output for the CustomField api to accept and return the old field types. This is essential for 3rd party integrations like `webform_civicrm`.
This compatibility mode can be turned off in APIv3 by passing `legacy_html_type => FALSE`.
APIv4 does not have a legacy mode as it's still fairly new.

### CODE REVIEW

**Comment by colemanw** [comment]
@pradpnayak this is where it's all heading...

**Comment by colemanw** [comment]
@eileenmcnaughton @pradpnayak this is now passing tests!

**Comment by colemanw** [comment]
@totten as soon as 5.26 is branched I can rebase this. If we merge it this week, then it'll have a month in `master` and then another month in RC.

**Comment by colemanw** [comment]
@totten @pradpnayak this is now rebased for 5.27

File: CRM/Custom/Form/Field.php [hunk 1]
```diff
@@ -950,6 +950,17 @@ public function postProcess() {
       $params['is_search_range'] = 0;
     }
 
+    // Serialization cannot be changed on update
+    if ($this->_id) {
+      unset($params['serialize']);
+    }
+    elseif (in_array($params['html_type'], ['Select', 'Country', 'StateProvince'])) {
```
Comment (pradpnayak) [inline CRM/Custom/Form/Field.php:957]: Shouldn't 
Country -> Select Country
StateProvince -> Select State/Province 

As per https://github.com/civicrm/civicrm-core/pull/16992/files#diff-cac3f19d5b4be8233ee77c257538bc40R163?

**Comment by pradpnayak** [comment]
Hi @colemanw 

I did some testing today most of the things seems to be working however there are some issue

1. Custom fields for multi state and multi country are not rendered as multi select
2. Multi-select field are not set by default on Contribution, Event and profile page.

**Comment by pradpnayak** [comment]
Notice error on create profile mode

Warning: in_array() expects parameter 2 to be array, null given in CRM_Core_BAO_CustomField::setProfileDefaults() (line 1251 of /Users/pradeep/Sites/civicrm/sites/all/modules/civicrm/CRM/Core/BAO/CustomField.php).

**Comment by colemanw** [comment]
@pradpnayak I'm unable to reproduce `#1` - on what screen are they appearing as single-select?

**Comment by colemanw** [comment]
retest this please

**Comment by colemanw** [comment]
retest this please

**Comment by colemanw** [comment]
retest this please

**Comment by pradpnayak** [comment]
Hi @colemanw 

#2 and #3 is fixed now

For #1 i am adding gif images 
Multistate


Found another issue but could replicate on older version also
When doing search contact using Adv search with filter using auto complete custom fields selecting more than one option it returns 0 results

Screen Shot 2020-05-12 at 23 28 43

File: CRM/Custom/Form/Field.php (hunk 1 above)
Comment (colemanw) [reply CRM/Custom/Form/Field.php:957]: You would think so, but this form is very strange. I still haven't figured out why it works at all.

Comment (colemanw) [reply CRM/Custom/Form/Field.php]: Ok I stepped through this with a debugger and you're right. I've updated the commit to check for any string starting with "Select".

**Comment by colemanw** [comment]
Thanks @pradpnayak I've now fixed `#1` and rebased.

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by colemanw** [comment]
Jenkins re test this please

**Comment by colemanw** [comment]
Jenkins re test this please

**Comment by colemanw** [comment]
Jenkins re test this please

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by colemanw** [comment]
Jenkins re test this please

**Comment by pradpnayak** [comment]
Hi @colemanw 

It seems to work fine now for all custom field types. Tested via Contact summary, Profile, Contribution page, event page, Contact api create/update.

**Comment by colemanw** [comment]
Great. We did it!

**Pull request merged**